### PR TITLE
feat: add keepalived to Anvil

### PR DESCRIPTION
**Pull request by SK1Y101**
Add the Keepalived charm to MAAS Anvil.

Allow specifying the virtual IP

addresses #10

### CODE REVIEW

File: anvil-python/anvil/jobs/common.py [hunk 1]
```diff
@@ -78,3 +79,14 @@ def validate_roles(
         return [Role[role.upper()] for role in value]
     except KeyError as e:
         raise click.BadParameter(str(e))
+
+
+def validate_ip_address(
```
Comment (skatsaounis) [inline anvil-python/anvil/jobs/common.py]: Maybe we need to remove that too

File: anvil-python/anvil/provider/local/commands.py [hunk 2]
```diff
@@ -95,6 +95,7 @@
 from anvil.jobs.common import (
     Role,
     roles_to_str_list,
+    validate_ip_address,
```
Comment (skatsaounis) [inline anvil-python/anvil/provider/local/commands.py]: same here for all the changes in `commands.py`

File: cloud/etc/deploy-haproxy/main.tf [hunk 3]
```diff
@@ -44,3 +44,35 @@ resource "juju_application" "haproxy" {
...
+  application {
+    name     = juju_application.haproxy.name
+    endpoint = "juju-info"
+  }
+
+  application {
+    name     = juju_application.keepalived.name
+    endpoint = "juju-info"
+  }
+}
```
Comment (skatsaounis) [inline cloud/etc/deploy-haproxy/main.tf:78]: nit: what about a new line in the end of the file?

File: cloud/etc/deploy-haproxy/variables.tf [hunk 4]
```diff
@@ -41,3 +41,21 @@ variable "machine_model" {
...
+  description = "Operator channel revision for Keepalived"
+  type        = number
+  default     = null
+}
+
+variable "virtual_ip" {
+  description = "Virtual IP Address for keepalived"
+  type        = string
+  default     = ""
+}
```
Comment (skatsaounis) [inline cloud/etc/deploy-haproxy/variables.tf:61]: nit: what about a new line in the end of the file?

File: anvil-python/anvil/commands/haproxy.py [hunk 5]
```diff
@@ -54,10 +64,34 @@ def __init__(
...
     def get_application_timeout(self) -> int:
         return HAPROXY_APP_TIMEOUT
 
+    def has_prompts(self) -> bool:
+        return True
+
+    def prompt(self, console: Console | None = None) -> None:
+        bootstrap_bank = QuestionBank(
+            questions={
+                "virtual_ip": PromptQuestion(
```
Comment (skatsaounis) [inline anvil-python/anvil/commands/haproxy.py:78]: what about adding the `validation_function` here? https://github.com/canonical/snap-openstack/blob/anvil-0.1/sunbeam-python/sunbeam/jobs/questions.py#L105

File: anvil-python/anvil/jobs/common.py (hunk 1 above)
Comment (skatsaounis) [reply anvil-python/anvil/jobs/common.py]: Or repurpose it to be a question validation function as suggested below

**Review by skatsaounis — CHANGES_REQUESTED** [review]
Great job @SK1Y101. I think that we need to update the README.md part that is referencing the VIP addition. After that, we are good to merge :tada:

File: README.md [hunk 6]
```diff
@@ -24,6 +24,8 @@ ubuntu@infra{1,2,3}:~$ newgrp snap_daemon
 ```
 
 ### Bootstrap the first node
+During installation of the HAProxy charm, you will be asked for a virtual_ip. Pass an empty value to disable it.
```
Comment (skatsaounis) [inline README.md:27]: If `accept_defaults` is used, do we still receive the question?

Comment (skatsaounis) [inline README.md:27]:
```suggestion

During installation of the HAProxy charm, you will be asked for a `virtual_ip`. Pass an empty value to disable it.
```

File: README.md [hunk 7]
```diff
@@ -24,6 +24,8 @@ ubuntu@infra{1,2,3}:~$ newgrp snap_daemon
...
+Passing any other valid IP will install the KeepAliveD charm, allowing you to connect to HA MAAS using the VIP.
```
Comment (skatsaounis) [inline README.md:28]:
```suggestion
Passing any other valid IP will install the Keepalived charm, allowing you to connect to HA MAAS using the VIP.
```

File: README.md [hunk 8]
```diff
@@ -42,6 +44,8 @@ Token for the Node infra3.: eyJuYW1lIjoibWFhcy00Lm1hYXMiLCJzZWNyZXQiOiI3MmE51234
 ```
 
 ### Join new nodes to the MAAS cluster
+During installation of the HAProxy charm, you will be asked for a virtual_ip. Pass an empty value to disable it.
```
Comment (skatsaounis) [inline README.md:47]:
```suggestion

During installation of the HAProxy charm, you will be asked for a `virtual_ip`. Pass an empty value to disable it.
```

File: README.md [hunk 9]
```diff
@@ -42,6 +44,8 @@ Token for the Node infra3.: eyJuYW1lIjoibWFhcy00Lm1hYXMiLCJzZWNyZXQiOiI3MmE51234
...
+Passing any other valid IP will install the KeepAliveD charm, allowing you to connect to HA MAAS using the VIP.
```
Comment (skatsaounis) [inline README.md:48]:
```suggestion
Passing any other valid IP will install the Keepalived charm, allowing you to connect to HA MAAS using the VIP.
```

File: README.md (hunk 8 above)
Comment (skatsaounis) [reply README.md:47]: Maybe we need to investigate a bit to make this question not been asked on joining new nodes.

**Review by skatsaounis — CHANGES_REQUESTED** [review]

File: anvil-python/anvil/commands/haproxy.py [hunk 10]
```diff
@@ -76,6 +81,8 @@ def has_prompts(self) -> bool:
         return True
 
     def prompt(self, console: Console | None = None) -> None:
+        previous_answers = load_answers(self.client, CONFIG_KEY)
```
Comment (SK1Y101) [inline anvil-python/anvil/commands/haproxy.py:84]: In theory this should fix #27

**Comment by SK1Y101** [comment]
@skatsaounis do these changes work for you?

**Review by skatsaounis — APPROVED** [review]
LGTM with today's changes. It is also passing my local tests successfully. I can set the VIP through the CLI prompt or a manifest. Or I can proceed without specifying a VIP. When I join another haproxy node I also have the option to accept-defaults or remember the previous answer.

@SK1Y101 could you also please take a look at the changes and then merge or update any needed change? Thank you :slightly_smiling_face:

**Comment by SK1Y101** [comment]
@skatsaounis the changes look good to me, if it passes the functional tests and you can see the expected behavior in testing I'm happy!

**Pull request merged**